core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java

```diff
@@ -985,54 +955,43 @@ private MutationSet sendMutationsToTabletServer(String location,
         throw new IOException(e);
       }
     }
+
+    private void close() {
+      this.closed = true;
+      binningThreadPool.shutdownNow();
+      sendThreadPool.shutdownNow();
+    }
   }
 
   // END code for sending mutations to tablet servers using background threads
 
-  private static class MutationSet {
+  private static class MutationSet extends ConcurrentHashMap<TableId,Queue<Mutation>> {
+
+    private static final long serialVersionUID = 1L;
 
-    private final HashMap<TableId,List<Mutation>> mutations;
     private long memoryUsed = 0;
 
-    MutationSet() {
-      mutations = new HashMap<>();
+    public void add(TableId table, Mutation mutation) {
+      memoryUsed += mutation.estimatedMemoryUsed();
```

Comment:
This add has a race condition and is not correct w/o sync.  This code may have relied on being called in a sync block when the obj was used by multiple threads (some mutation set objs are only used by a single thread within a method).

Reply:
I sync'd this method